### PR TITLE
add messaging queue implementation

### PR DESCRIPTION
**Pull request by KyleTTownsend**

### CODE REVIEW

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 1]
```diff
@@ -18,16 +21,373 @@ public class UserController : BaseController
...
+                if (user == null)
+                {
+                    logger.LogError("User model was null");
+                    msg.Label = "400";
+                    msg.Body = Encoding.UTF8.GetBytes("User model was null.");
+                    return msg;
+                }
+                else
+                {
+                    var id = user.UserId;
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove this unused 'id' local variable. ([csharpsquid:S1481](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1481))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCE&open=AWVoarn6elKyZq-p4oCE)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCE

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 2]
```diff
@@ -18,16 +21,373 @@ public class UserController : BaseController
...
+            catch (Exception e)
+            {
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
 
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 3]
```diff
@@ -18,16 +21,365 @@ public class UserController : BaseController
...
+            catch (Exception e)
+            {
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 4]
```diff
@@ -18,16 +21,365 @@ public class UserController : BaseController
...
+                if (user == null)
+                {
+                    logger.LogError("User model was null");
+                    msg.Label = "400";
+                    msg.Body = Encoding.UTF8.GetBytes("User model was null.");
+                    return msg;
+                }
+                else
+                {
+                    var id = user.UserId;
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove this unused 'id' local variable. ([csharpsquid:S1481](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1481))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCE&open=AWVoarn6elKyZq-p4oCE)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCE

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 5]
```diff
@@ -18,16 +20,368 @@ public class UserController : BaseController
...
+            catch (Exception e)
+            {
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 6]
```diff
@@ -18,16 +20,368 @@ public class UserController : BaseController
...
+                if (user == null)
+                {
+                    logger.LogError("User model was null");
+                    msg.Label = "400";
+                    msg.Body = Encoding.UTF8.GetBytes("User model was null.");
+                    return msg;
+                }
+                else
+                {
+                    var id = user.UserId;
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove this unused 'id' local variable. ([csharpsquid:S1481](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1481))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCE&open=AWVoarn6elKyZq-p4oCE)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCE

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 7]
```diff
@@ -18,16 +20,368 @@ public class UserController : BaseController
...
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove this unused 'id' local variable. ([csharpsquid:S1481](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1481))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCE&open=AWVoarn6elKyZq-p4oCE)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCE

File: src/ServiceHub.User.Service/Controllers/UserController.cs (hunk 5 above)
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 8]
```diff
@@ -18,16 +20,376 @@ public class UserController : BaseController
...
+            {
+                msg = new Message();
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 9]
```diff
@@ -18,16 +20,376 @@ public class UserController : BaseController
...
+                if (user == null)
+                {
+                    logger.LogError("User model was null");
+                    msg.Label = "400";
+                    msg.Body = Encoding.UTF8.GetBytes("User model was null.");
+                    return msg;
+                }
+                else
+                {
+                    var id = user.UserId;
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove this unused 'id' local variable. ([csharpsquid:S1481](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1481))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCE&open=AWVoarn6elKyZq-p4oCE)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCE

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 10]
```diff
@@ -18,16 +20,394 @@ public class UserController : BaseController
...
+                    return null;
+                msg.CorrelationId = message.MessageId;
+                logger.LogError(e, "Exception thrown during Post");
+                msg.Body = Encoding.UTF8.GetBytes("Exception thrown during Post");
+                msg.Label = "500";
+                return msg;
+            }
+        }
+
+        public async Task<Message> PutQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVtVKHOXKENQn1SquHL&open=AWVtVKHOXKENQn1SquHL)

[ISSUE_KEY]: AWVtVKHOXKENQn1SquHL

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 11]
```diff
@@ -18,16 +20,394 @@ public class UserController : BaseController
...
+                if (user == null)
+                {
+                    logger.LogError("User model was null");
+                    msg.Label = "400";
+                    msg.Body = Encoding.UTF8.GetBytes("User model was null.");
+                    return msg;
+                }
+                else
+                {
+                    var id = user.UserId;
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove this unused 'id' local variable. ([csharpsquid:S1481](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS1481))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCE&open=AWVoarn6elKyZq-p4oCE)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCE

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 12]
```diff
@@ -18,16 +20,394 @@ public class UserController : BaseController
...
+                    return null;
+                msg.CorrelationId = message.MessageId;
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 13]
```diff
@@ -18,16 +20,392 @@ public class UserController : BaseController
...
+                    return null;
+                msg.CorrelationId = message.MessageId;
+                logger.LogError(e, "Exception thrown during Post");
+                msg.Body = Encoding.UTF8.GetBytes("Exception thrown during Post");
+                msg.Label = "500";
+                return msg;
+            }
+        }
+
+        public async Task<Message> PutQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVtVKHOXKENQn1SquHL&open=AWVtVKHOXKENQn1SquHL)

[ISSUE_KEY]: AWVtVKHOXKENQn1SquHL

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 14]
```diff
@@ -18,16 +20,392 @@ public class UserController : BaseController
...
+                    return null;
+                msg.CorrelationId = message.MessageId;
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 18 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs (hunk 13 above)
Comment (Hoagithor) [reply src/ServiceHub.User.Service/Controllers/UserController.cs]: Unavoidable while duplicating existing implementation from HTTP.

File: src/ServiceHub.User.Service/Controllers/UserController.cs (hunk 14 above)
Comment (Hoagithor) [reply src/ServiceHub.User.Service/Controllers/UserController.cs]: Unavoidable while duplicating existing implementation from HTTP.

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 15]
```diff
@@ -18,16 +20,380 @@ public class UserController : BaseController
...
+                msg = new Message();
+                msg.CorrelationId = message.MessageId;
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 16]
```diff
@@ -27,10 +33,370 @@ public class UserController : BaseController
...
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 17]
```diff
@@ -19,6 +21,10 @@ public class UserController : BaseController
     {
         private readonly IUserRepository _userRepository;
         private readonly IQueueFactory _queues;
+        private readonly IQueueClient queueSender;
+        private readonly IQueueClient queueReceiver;
+        private static readonly string sendQueueName = "usertohs";
+        private static readonly string receiveQueueName = "hstouser";
         private readonly LastModifiedTracker lmt;
```
Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Remove unassigned field 'lmt', or set its value. ([csharpsquid:S3459](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3459))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWV7mIADX2sY87p0jiwz&open=AWV7mIADX2sY87p0jiwz)

[ISSUE_KEY]: AWV7mIADX2sY87p0jiwz

Comment (Hoagithor) [reply src/ServiceHub.User.Service/Controllers/UserController.cs]: Added from broker merge, left unchanged in accordance.

File: src/ServiceHub.User.Service/Controllers/UserController.cs (hunk 16 above)
Comment (Hoagithor) [reply src/ServiceHub.User.Service/Controllers/UserController.cs]: Unavoidable due to existing method whose implementation we copied having this complexity.

File: src/ServiceHub.User.Service/Controllers/UserController.cs [hunk 18]
```diff
@@ -27,10 +31,374 @@ public class UserController : BaseController
...
+                msg = new Message();
+                msg.CorrelationId = message.MessageId;
+                logger.LogError(e, "Resource was not found in database.");
+                msg.Label = "404";
+                msg.Body = Encoding.UTF8.GetBytes("Not Found");
+                return msg;
+            }
+        }
+
+        public async Task<Message> GetByGenderQ(Message message)
```
[the remark this answers is not in the document]
Comment (Hoagithor) [reply src/ServiceHub.User.Service/Controllers/UserController.cs]: Unavoidable due to mimicking previous implementation.

Comment (fredbelotte) [inline src/ServiceHub.User.Service/Controllers/UserController.cs]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. ([csharpsquid:S3776](https://sonarcloud.io/organizations/revaturecloud/rules#rule_key=csharpsquid%3AS3776))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=servicehub%3Auser&pullRequest=95&issues=AWVoarn6elKyZq-p4oCD&open=AWVoarn6elKyZq-p4oCD)

[ISSUE_KEY]: AWVoarn6elKyZq-p4oCD

**Pull request merged**